text/0000-bench-utils.md

```diff
@@ -0,0 +1,207 @@
+- Feature Name: black_box
+- Start Date: 2018-03-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds one function, `core::hint::black_box`, which is a hint to the
+optimizer to disable certain compiler optimizations. 
+
+# Motivation
+[motivation]: #motivation
+
+A tool for preventing compiler optimizations is widely useful. One application
+is writing synthetic benchmarks, where, due to the constrained nature of the
+benchmark, the compiler is able to perform optimizations that wouldn't otherwise
+trigger in practice. Another application is writing constant time code, where it
+is undesirable for the compiler to optimize certain operations depending on the
+context in which they are executed.
+
+The implementation of this function is backend-specific and currently requires
+inline assembly. No viable alternative is available in stable Rust.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `hint::black_box`
+
+The function:
+
+```rust
+/// An _unknown_ function that returns `x`.
+pub unsafe fn black_box<T>(x: T) -> T;
+```
+
+is an _unknown_ function, that is, a function that the compiler cannot make any
+assumptions about. It can potentially use `x` in any possible valid way that
+`unsafe` Rust code is allowed to, and requires the compiler to be maximally
```

Comment:
> The emphasis here is on allowed to -- as opposed to "any possible way that unsafe Rust could use x". I think this should be made clearer. Maybe link to the nomicon?

Thanks, I think this makes more sense.

> Actually, "any possible way that safe Rust could use x" is almost more precise, but may also give the wrong impression.

How so? I thought unsafe Rust was allowed to use unsafe for, e.g., dereferencing raw pointers, while if this would be a safe function, it couldn't do so. In any case, whatever this function does, it cannot invoke undefined behavior and the compiler is allowed to assume that.